File: MiniDOMOffice/CRM/Calendar/CRMCalProvider.vb
Imports minidom
Imports minidom.Sistema
Imports minidom.Databases

Imports minidom.Anagrafica

Partial Public Class CustomerCalls




    Public Class CRMCalProvider
        Inherits BaseCalendarActivitiesProvider

        Private Class FaxInfo
            Implements IComparable

            Public ID As Integer
            Public PO As Integer

            Public Sub New(ByVal c As FaxDocument)
                Me.ID = GetID(c)
                Me.PO = c.IDPuntoOperativo
            End Sub

            Public Sub New(ByVal id As Integer)
                Me.ID = id
                Me.PO = 0
            End Sub

            Private Function CompareTo(obj As Object) As Integer Implements IComparable.CompareTo
                Dim tmp As FaxInfo = obj
                Return Arrays.Compare(Me.ID, tmp.ID)
            End Function
        End Class

        Private listLock As New Object
        Private m_FaxList As FaxInfo()

        Public Sub New()
            AddHandler CustomerCalls.CRM.ContattoModificato, AddressOf handleContatto
            AddHandler CustomerCalls.CRM.ContattoEliminato, AddressOf handleContatto
            AddHandler CustomerCalls.CRM.NuovoContatto, AddressOf handleContatto
            Me.m_FaxList = Nothing
        End Sub

        Private Function IsInFaxList(ByVal id As Integer) As Boolean
            Return Arrays.Len(Me.m_FaxList) > 0 AndAlso Arrays.BinarySearch(Me.m_FaxList, New FaxInfo(id)) >= 0
        End Function

        Private Sub AddToFaxList(ByVal c As FaxDocument)
            Dim i As Integer = -1
            Dim info As New FaxInfo(c)
            If (Arrays.Len(Me.m_FaxList) > 0) Then i = Arrays.BinarySearch(Me.m_FaxList, info)
            If (i >= 0) Then
                Me.m_FaxList(i) = info
            Else
                Me.m_FaxList = Arrays.Push(Me.m_FaxList, info)
                Array.Sort(Me.m_FaxList)
            End If
        End Sub

        Private Sub RemoveFromFaxList(ByVal id As Integer)
            Dim i As Integer = -1
            If (Arrays.Len(Me.m_FaxList) > 0) Then i = Arrays.BinarySearch(Me.m_FaxList, New FaxInfo(id))
            If (i >= 0) Then Me.m_FaxList = Arrays.RemoveAt(Me.m_FaxList, i)
        End Sub

        Private ReadOnly Property FaxList As CCollection(Of FaxInfo)
            Get
                SyncLock Me.listLock
                    If (Me.m_FaxList Is Nothing) Then
                        Dim cursor1 As New CustomerCalls.CCustomerCallsCursor
                        Dim tmp As New System.Collections.ArrayList
                        cursor1.Stato.Value = ObjectStatus.OBJECT_VALID
                        cursor1.IDPersona.Value = 0
                        cursor1.ClassName.Value = "FaxDocument"
                        cursor1.IgnoreRights = True
                        cursor1.ID.SortOrder = SortEnum.SORT_ASC
                        While Not cursor1.EOF
                            tmp.Add(New FaxInfo(cursor1.Item))
                            cursor1.MoveNext()
                        End While
                        cursor1.Dispose()
                        Me.m_FaxList = tmp.ToArray(GetType(FaxInfo))
                        'If (Me.m_FaxList IsNot Nothing) Then Array.Sort(Me.m_FaxList)
                    End If
                    Dim ret As New CCollection(Of FaxInfo)
                    If (Me.m_FaxList IsNot Nothing) Then ret.AddRange(Me.m_FaxList)
                    Return ret
                End SyncLock
            End Get
        End Property

        Private Sub handleContatto(ByVal e As ContattoEventArgs)
            Dim c As CContattoUtente = e.Contatto
            If (TypeOf (c) Is FaxDocument) Then
                SyncLock Me.listLock
                    If (c.Stato = ObjectStatus.OBJECT_VALID AndAlso c.IDPersona = 0) Then
                        Me.AddToFaxList(c)
                    Else
                        Me.RemoveFromFaxList(GetID(c))
                    End If
                End SyncLock
            End If
        End Sub


        Public Overrides Sub DeleteActivity(item As ICalendarActivity, Optional force As Boolean = False)

        End Sub

        Public Overrides Function GetActivePersons(nomeLista As String, fromDate As Date?, toDate As Date?, Optional ufficio As Integer = 0, Optional operatore As Integer = 0) As CCollection(Of CActivePerson)
            Return New CCollection(Of CActivePerson)
        End Function

        Public Overrides Function GetPendingActivities() As CCollection(Of ICalendarActivity)
            Return New CCollection(Of ICalendarActivity)
        End Function

        Public Overrides Function GetScadenze(fromDate As Date?, toDate As Date?) As CCollection(Of ICalendarActivity)
            Return New CCollection(Of ICalendarActivity)
        End Function

        Private Function GetAllowedFaxList() As CCollection(Of FaxInfo)
            Dim u As CUser = Sistema.Users.CurrentUser
            Dim ret As New CCollection(Of FaxInfo)
            For Each f As FaxInfo In Me.FaxList
                If (f.PO = 0) OrElse (u.Uffici.HasOffice(f.PO)) Then
                    ret.Add(f)
                End If
            Next
            Return ret
        End Function

        Public Overrides Function GetToDoList(user As CUser) As CCollection(Of ICalendarActivity)
            SyncLock Me.listLock
                Dim ret As New CCollection(Of ICalendarActivity)
                Dim act As CCalendarActivity

                If (CRM.CRMGroup.Members.Contains(Sistema.Users.CurrentUser)) Then
                    Dim items As CCollection(Of FaxInfo) = Me.GetAllowedFaxList
                    If items.Count > 0 Then
                        act = New CCalendarActivity
                        DirectCast(act, ICalendarActivity).SetProvider(Me)
                        act.Descrizione = "Ci sono " & items.Count & " fax da associare"
                        act.DataInizio = DateUtils.Now
                        act.GiornataIntera = True
                        act.Categoria = "Normale"
                        act.Flags = CalendarActivityFlags.IsAction
                        act.IconURL = "/widgets/images/activities/faxdaassociare.png"
                        act.Note = "CRMShowFaxNonAssociati"
                        act.Stato = ObjectStatus.OBJECT_VALID
                        ret.Add(act)
                    End If
                End If
                Return ret
            End SyncLock
        End Function

        Public Overrides Sub SaveActivity(item As ICalendarActivity, Optional force As Boolean = False)

        End Sub

        Public Overrides ReadOnly Property UniqueName As String
            Get
                Return "CRMCALPROVIDER"
            End Get
        End Property
    End Class

End Class

File: MiniDOMOffice/CRM/Interazioni/FaxDocuments/CRMProvider/CRMCalProvider.cs
using System;
using System.Collections;
using System.Collections.Generic;
using DMD;
using DMD.XML;
using DMD.Databases;
using DMD.Databases.Collections;
using minidom;
using minidom.repositories;
using static minidom.Sistema;
using static minidom.Anagrafica;
using static minidom.CustomerCalls;

namespace minidom
{
    public partial class CustomerCalls
    {

        /// <summary>
        /// Provider di eventi del calendario
        /// </summary>
        public class CRMCalProvider 
            : Sistema.BaseCalendarActivitiesProvider
        {

            /// <summary>
            /// Informazioni sui fax in attesa
            /// </summary>
            private class FaxInfo 
                : IComparable, IComparable<FaxInfo>
            {
                /// <summary>
                /// ID del fax
                /// </summary>
                public int ID;

                /// <summary>
                /// ID del punto operativo
                /// </summary>
                public int PO;

                /// <summary>
                /// Costruttore
                /// </summary>
                /// <param name="c"></param>
                public FaxInfo(FaxDocument c)
                {
                    ID = DBUtils.GetID(c, 0);
                    PO = c.IDPuntoOperativo;
                }

                /// <summary>
                /// Costruttore
                /// </summary>
                /// <param name="id"></param>
                public FaxInfo(int id)
                {
                    ID = id;
                    PO = 0;
                }

                /// <summary>
                /// Compara due oggetti
                /// </summary>
                /// <param name="tmp"></param>
                /// <returns></returns>
                public int CompareTo(FaxInfo tmp)
                {
                    return DMD.Arrays.Compare(this.ID, tmp.ID);
                }

                int IComparable.CompareTo(object obj)
                {
                    return this.CompareTo((FaxInfo)obj);
                }
            }

            private object listLock = new object();
            private FaxInfo[] m_FaxList;

            /// <summary>
            /// Costruttore
            /// </summary>
            public CRMCalProvider()
            {
                minidom.CustomerCalls.CRM.ItemModified += handleContatto;
                minidom.CustomerCalls.CRM.ItemCreated += handleContatto;
                minidom.CustomerCalls.CRM.ItemDeleted += handleContatto;
                m_FaxList = null;
            }

            private bool IsInFaxList(int id)
            {
                int localBinarySearch() { var argitem = new FaxInfo(id); var ret = DMD.Arrays.BinarySearch(m_FaxList, argitem); return ret; }

                return DMD.Arrays.Len(m_FaxList) > 0 && localBinarySearch() >= 0;
            }

            private void AddToFaxList(FaxDocument c)
            {
                int i = -1;
                var info = new FaxInfo(c);
                if (DMD.Arrays.Len(m_FaxList) > 0)
                    i = DMD.Arrays.BinarySearch(m_FaxList, info);

                if (i >= 0)
                {
                    m_FaxList[i] = info;
                }
                else
                {
                    m_FaxList = DMD.Arrays.Append(m_FaxList, info);
                    DMD.Arrays.Sort(m_FaxList);
                }
            }

            private void RemoveFromFaxList(int id)
            {
                int i = -1;
                if (DMD.Arrays.Len(m_FaxList) > 0)
                {
                    var argitem = new FaxInfo(id);
                    i = DMD.Arrays.BinarySearch(m_FaxList, argitem);
                }

                if (i >= 0)
                    m_FaxList = DMD.Arrays.RemoveAt(m_FaxList, i);
            }

            private CCollection<FaxInfo> FaxList
            {
                get
                {
                    lock (listLock)
                    {
                        if (m_FaxList is null)
                        {
                            var tmp = new List<FaxInfo>();
                            using (var cursor1 = new CContattoUtenteCursor())
                            {
                                cursor1.Stato.Value = ObjectStatus.OBJECT_VALID;
                                cursor1.IDPersona.Value = 0;
                                cursor1.ClassName.Value = "FaxDocument";
                                cursor1.IgnoreRights = true;
                                cursor1.ID.SortOrder = SortEnum.SORT_ASC;
                                while (cursor1.Read())
                                {
                                    tmp.Add(new FaxInfo((FaxDocument)cursor1.Item));
                                }

                            }
                            m_FaxList = tmp.ToArray();
                            // If (Me.m_FaxList IsNot Nothing) Then Array.Sort(Me.m_FaxList)
                        }

                        var ret = new CCollection<FaxInfo>();
                        if (m_FaxList is object)
                            ret.AddRange(m_FaxList);
                        return ret;
                    }
                }
            }

            private void handleContatto(object sender, ItemEventArgs<CContattoUtente> e)
            {
                var c = e.Item;
                if (c is FaxDocument)
                {
                    lock (listLock)
                    {
                        if (c.Stato == ObjectStatus.OBJECT_VALID && c.IDPersona == 0)
                        {
                            AddToFaxList((FaxDocument)c);
                        }
                        else
                        {
                            RemoveFromFaxList(DBUtils.GetID(c, 0));
                        }
                    }
                }
            }

            /// <summary>
            /// Elimina l'attività
            /// </summary>
            /// <param name="item"></param>
            /// <param name="force"></param>
            public override void DeleteActivity(Sistema.ICalendarActivity item, bool force = false)
            {
                //TODO DeleteActivity
                throw new NotImplementedException();
            }

            /// <summary>
            /// Restituisce le persone attive
            /// </summary>
            /// <param name="nomeLista"></param>
            /// <param name="fromDate"></param>
            /// <param name="toDate"></param>
            /// <param name="ufficio"></param>
            /// <param name="operatore"></param>
            /// <returns></returns>
            public override CCollection<Sistema.CActivePerson> GetActivePersons(string nomeLista, DateTime? fromDate, DateTime? toDate, int ufficio = 0, int operatore = 0)
            {
                //TODO GetActivePersons
                return new CCollection<Sistema.CActivePerson>();
            }

            /// <summary>
            /// Restituisce le attività in sospeso
            /// </summary>
            /// <returns></returns>
            public override CCollection<Sistema.ICalendarActivity> GetPendingActivities()
            {
                //TODO GetPendingActivities
                return new CCollection<Sistema.ICalendarActivity>();
            }

            /// <summary>
            /// Restituisce le scadenze
            /// </summary>
            /// <param name="fromDate"></param>
            /// <param name="toDate"></param>
            /// <returns></returns>
            public override CCollection<Sistema.ICalendarActivity> GetScadenze(DateTime? fromDate, DateTime? toDate)
            {
                //TODO GetScadenze
                return new CCollection<Sistema.ICalendarActivity>();
            }

            private CCollection<FaxInfo> GetAllowedFaxList()
            {
                lock(this.listLock)
                {
                    var u = Sistema.Users.CurrentUser;
                    var ret = new CCollection<FaxInfo>();
                    foreach (FaxInfo f in FaxList)
                    {
                        if (f.PO == 0 || u.Uffici.HasOffice(f.PO))
                        {
                            ret.Add(f);
                        }
                    }

                    return ret;
                }
            }

            /// <summary>
            /// Restituisce l'elenco delle cose da fare
            /// </summary>
            /// <param name="user"></param>
            /// <returns></returns>
            public override CCollection<Sistema.ICalendarActivity> GetToDoList(Sistema.CUser user)
            {
                var ret = new CCollection<Sistema.ICalendarActivity>();
                Sistema.CCalendarActivity act;
                if (CRM.CRMGroup.Members.Contains(Sistema.Users.CurrentUser))
                {
                    var items = GetAllowedFaxList();
                    if (items.Count > 0)
                    {
                        act = new Sistema.CCalendarActivity();
                        ((Sistema.ICalendarActivity)act).SetProvider(this);
                        act.Descrizione = "Ci sono " + items.Count + " fax da associare";
                        act.DataInizio = DMD.DateUtils.Now();
                        act.GiornataIntera = true;
                        act.Categoria = "Normale";
                        act.Flags = Sistema.CalendarActivityFlags.IsAction;
                        act.IconURL = "/widgets/images/activities/faxdaassociare.png";
                        act.Note = "CRMShowFaxNonAssociati";
                        act.Stato = ObjectStatus.OBJECT_VALID;
                        ret.Add(act);
                    }
                }

                return ret;
            }

            /// <summary>
            /// Salva l'attività
            /// </summary>
            /// <param name="item"></param>
            /// <param name="force"></param>
            public override void SaveActivity(Sistema.ICalendarActivity item, bool force = false)
            {
                //TODO SaveActivity
                throw new NotImplementedException();
            }

            /// <summary>
            /// Restituisce il nome univoco dell'handler
            /// </summary>
            public override string UniqueName
            {
                get
                {
                    return "CRMCALPROVIDER";
                }
            }
        }
    }
}